### PR TITLE
Add FreeRTOS Kernel and C++ Support to Simulator

### PR DESCRIPTION
**Pull request by MootSeeker**
This pull request introduces several key enhancements to the project:

- FreeRTOS Integration: Added the FreeRTOS kernel to the simulator, enabling developers to fully simulate an embedded system environment, including task scheduling and real-time operations.
- LVGL Task Creation: Refactored the LVGL execution to run within a FreeRTOS task, providing a more realistic simulation of how LVGL would operate on an actual device.
- C++ Transition: Converted the project from C to C++ to leverage modern programming features and improve code structure and readability.

Changes:

- Added FreeRTOS as a submodule.
- Created freertos_config target to manage FreeRTOS configuration.
- Refactored main.c to main.cpp with appropriate C++ syntax and constructs.
- Implemented memory management and error handling hooks for FreeRTOS in C++.

Benefits:

- Improved simulation accuracy of embedded systems.
- Enhanced maintainability and flexibility of the codebase through C++.
- Full task management and scheduling emulation, closely reflecting real device behavior.

Please review the changes and provide feedback.

### CODE REVIEW

**Review by kisvegabor — COMMENTED** [review]
It's great update! See my comments below. Please also mention this new option in the README.

**Comment by MootSeeker** [comment]
@kisvegabor thank you very much for your feedback, I'm happy to hear that. I will be looking at your points and adapting the project structure so that it is compatible with the original master.

File: main/src/main.cpp [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+ * @return 	Always returns 0
+ */
+int main(int argc, char **argv)
+{
+    (void)argc; /*Unused*/
+    (void)argv; /*Unused*/
+
+    printf("Start FreeRTOS\n");
+
+    xTaskCreate( lvgl_task,
```
[the remark this answers is not in the document]
Comment (MootSeeker) [reply main/src/main.cpp]: Where I can find the documentation to `lv_thread_create`? 
When I set the `#define LV_USE_OS   LV_OS_NONE` to `LV_OS_FREERTOS` the whole application is crashing when enting the `lv_timer_handler();` function.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,14 +1,49 @@
...
+
+# SquareLineStudio UI files
+include_directories(${PROJECT_SOURCE_DIR}/UI)
+file(GLOB UI_SOURCES 
+    "${PROJECT_SOURCE_DIR}/UI/*.c" 
+    "${PROJECT_SOURCE_DIR}/UI/*.h" 
+    "${PROJECT_SOURCE_DIR}/UI/screens/ui_Screen1.c"
+    "${PROJECT_SOURCE_DIR}/UI/ui.c"
+    "${PROJECT_SOURCE_DIR}/UI/components/ui_comp_hook.c"
+    "${PROJECT_SOURCE_DIR}/UI/ui_helpers.c"
```
[the remark this answers is not in the document]
Comment (MootSeeker) [reply CMakeLists.txt]: removed

File: main/src/freertos_main.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+#include "lvgl/src/osal/lv_os.h"
+
+#if LV_USE_OS == LV_OS_FREERTOS
+
+#include "lvgl/examples/lv_examples.h"
+#include <cstdio>  // For printf in C++
+
+// Task identifiers used for managing FreeRTOS tasks
+enum TaskIdentifier_t {
```
Comment (kisvegabor) [inline main/src/freertos_main.cpp:14]: Is it used somewhere?

File: main/src/FreeRTOS_Posix_Port.c
Comment (kisvegabor) [inline main/src/FreeRTOS_Posix_Port.c:1]: The content of this file is not used in the project. Do we need it?

File: main/src/freertos_main.cpp [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+ * @param   None
+ * @return  None
+ */
+extern "C" void freertos_main()
+{
+    // Initialisiere LVGL und andere Ressourcen
+
+    // Erstelle den LVGL-Task
+    lv_thread_prio_t prio = static_cast<lv_thread_prio_t>(tskIDLE_PRIORITY + 1);
+    if (lv_thread_init(&task_handles.lvgl_thread, prio, lvgl_task, 16384, nullptr) != LV_RESULT_OK) {
```
Comment (kisvegabor) [inline main/src/freertos_main.cpp:169]: Let's use the FreeRTOS API here as well for the same of simplicity.

File: main/src/freertos_main.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        printf("Fehler beim Erstellen des LVGL-Tasks\n");
+        // Fehlerbehandlung
+    }
+
+    // Erstelle einen weiteren Task
+    if (xTaskCreate(another_task, "Another Task", 4096, nullptr, 1, &task_handles.another_task_handle) != pdPASS) {
```
Comment (kisvegabor) [inline main/src/freertos_main.cpp:175]:
```suggestion
    if (xTaskCreate(another_task, "Another Task", 4096, nullptr, 1, nullptr) != pdPASS) {
```

to eliminate `TaskHandles_t`.

File: main/src/freertos_main.cpp [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        printf("Fehler beim Erstellen eines weiteren Tasks\n"); 
+        // Fehlerbehandlung
+    }
+
+    // Starte den Scheduler
```
Comment (kisvegabor) [inline main/src/freertos_main.cpp:180]: For the sake of consistency please use ` /* ... */` comments.

File: README.md [hunk 7]
```diff
@@ -1,7 +1,19 @@
-# Simulator project for LVGL embedded GUI Library
+# Simulator project for LVGL embedded GUI Library running on FreeROS
```
Comment (kisvegabor) [inline README.md:1]:
```suggestion
# Simulator project for LVGL embedded GUI Library running on FreeRTOS
```

File: README.md [hunk 8]
```diff
@@ -1,7 +1,19 @@
...
+
+### What We've Accomplished:
+Integration of FreeRTOS: We added FreeRTOS as a submodule and configured the build system to include the necessary source files and headers.
+
+- Task Creation: A basic FreeRTOS task was implemented to initialize and run LVGL within the simulated environment.
+- Scheduler Integration: The FreeRTOS scheduler was incorporated to manage the execution of tasks, simulating a real-time environment.
+- LVGL Task Execution: We refactored the original code to run LVGL in a dedicated FreeRTOS task, allowing periodic handling through the FreeRTOS scheduling mechanism.
+
+This setup provides a more realistic development and testing environment, closely mirroring the conditions of an actual embedded system running FreeRTOS.
+
```
Comment (kisvegabor) [inline README.md:16]: It should describe what is in the repository and not the history. 

For example:
```md
## (RT)OS support

Works with any OS... Has build in support for FreeRTOS....
```

Below tell more about FreeRTOS (as you already did)

File: README.md [hunk 9]
```diff
@@ -108,3 +129,15 @@ git checkout release/6.0
 make
 sudo make install
 ```
+
+## Load GUI Design from SquareLine Studio
```
Comment (kisvegabor) [inline README.md:133]: This project is natural to 3rd party UI editors so please remove this section.

File: README.md [hunk 10]
```diff
@@ -39,6 +51,15 @@ sudo pacman -Syu && sudo pacman -S sdl2 libsdl2-devel sdl2_mixer sdl2-devel base
 
 ## Usage
 
+### FreeRTOS configuration
```
Comment (kisvegabor) [inline README.md:54]: Please move this section after `### Visual Studio Code` as setting up VSCode is the first step.

**Review by kisvegabor — COMMENTED** [review]
Thank you for the update! 

I add some comments, but they are quite simple.

File: main/src/FreeRTOS_Posix_Port.c
Comment (MootSeeker) [reply main/src/FreeRTOS_Posix_Port.c:1]: Correct, it is only used when you create an event. It was intended for the original project, but is not needed in this case.

File: main/src/freertos_main.cpp (hunk 3 above)
Comment (MootSeeker) [reply main/src/freertos_main.cpp:14]: No, can be removed.

File: main/src/freertos_main.cpp (hunk 4 above)
Comment (MootSeeker) [reply main/src/freertos_main.cpp:169]: Sure, I will change it back.

File: main/src/FreeRTOS_Posix_Port.c
Comment (MootSeeker) [reply main/src/FreeRTOS_Posix_Port.c:1]: I was a bit too fast with this answer this functions are used in heap4.c and port.c for example in `pxPortInitialiseStack()`

File: main/src/FreeRTOS_Posix_Port.c [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+#include <stdlib.h>
+
+/* Structure representing an event, consisting of a condition variable and a mutex */
+typedef struct Event
+{
+    pthread_cond_t cond;      /* Condition variable used to signal the event */
+    pthread_mutex_t mutex;    /* Mutex to protect access to the condition variable */
+} Event_t;
+
+// ........................................................................................................
```
Comment (kisvegabor) [inline main/src/FreeRTOS_Posix_Port.c:18]:
```suggestion
```

File: main/src/main.c [hunk 12]
```diff
@@ -80,6 +86,13 @@ int main(int argc, char **argv)
     usleep(5 * 1000);
   }
 
+  #elif LV_USE_OS == LV_OS_FREERTOS
+
+  // Run FreeRTOS and create lvgl task
```
Comment (kisvegabor) [inline main/src/main.c:91]:
```suggestion
  /* Run FreeRTOS and create lvgl task */
```

File: README.md [hunk 13]
```diff
@@ -1,15 +1,15 @@
-# Simulator project for LVGL embedded GUI Library
+# Simulator project for LVGL embedded GUI Library running on FreeRTOS
```
Comment (kisvegabor) [inline README.md:1]:
```suggestion
# VSCode Simulator project for LVGL

[LVGL](https://github.com/lvgl/lvgl) is written mainly for microcontrollers and embedded systems, however you can run the library **on your PC** as well without any embedded hardware. The code written on PC can be simply copied when your are using an embedded system.

This project is pre-configured for VSCode and should work work on Windows, Linux and MacOs as well. FreeRTOS is also included and can be optionally enabled to better simulate embedded system's behavior.  
```

File: README.md [hunk 14]
```diff
@@ -1,15 +1,15 @@
...
 
 The [LVGL](https://github.com/lvgl/lvgl) is written mainly for microcontrollers and embedded systems, however you can run the library **on your PC** as well without any embedded hardware. The code written on PC can be simply copied when your are using an embedded system.
 
-This project is configured for [VSCode](https://code.visualstudio.com) and only tested on Linux, although this may work on OSx or WSL. It requires a working version of GCC, GDB and make in your path.
+## (RT)OS support
+Works with any OS like pthred, Windows, FreeRTOS, etc. It has build in support for FreeRTOS.
```
Comment (kisvegabor) [inline README.md:6]: This section shouldn't be required here.

**Review by kisvegabor — COMMENTED** [review]
Thank you, just a final touch! :pray:

File: README.md (hunk 14 above)
Comment (MootSeeker) [reply README.md]: I removed this section and placed it at the end with a small hint.

**Comment by MootSeeker** [comment]
@kisvegabor do you close this PR?

File: README.md [hunk 15]
```diff
@@ -1,15 +1,12 @@
...
+[LVGL](https://github.com/lvgl/lvgl) is written mainly for microcontrollers and embedded systems, however you can run the library **on your PC** as well without any embedded hardware. The code written on PC can be simply copied when your are using an embedded system.
 
-To allow debugging inside VSCode you will also require a GDB [extension](https://marketplace.visualstudio.com/items?itemName=webfreak.debug) or other suitable debugger. All the requirements, build and debug settings have been pre-configured in the [.workspace](simulator.code-workspace) file.
+This project is pre-configured for VSCode and should work work on Windows, Linux and MacOs as well. FreeRTOS is also included and can be optionally enabled to better simulate embedded system's behavior.  
 
-The project can use **SDL** but it can be easily relaced by any other built-in LVGL dirvers.
+The [LVGL](https://github.com/lvgl/lvgl) is written mainly for microcontrollers and embedded systems, however you can run the library **on your PC** as well without any embedded hardware. The code written on PC can be simply copied when your are using an embedded system.
```
Comment (kisvegabor) [inline README.md:7]:
```suggestion
```

It was duplicated

**Pull request merged**